File: DataStructureAndAlgorithm/DataStructures/Trees/C++/BinarySearchTree.cpp
#include "BST_mine.h"
#include <iostream>

BST::BST() : root(nullptr) {}

BST::~BST() {
  deleteTree(root);
}

void BST::deleteTree(Node* node) {
  if (node != nullptr) {
    deleteTree(node->left);
    deleteTree(node->right);
    delete node;
  }
}

void BST::insert(int data) {
  root = insert(root, data);
}

BST::Node* BST::insert(Node* node, int data) {
  if (!node) return new Node(data);
  
  if (data < node->data) {
    node->left = insert(node->left, data);
  } else if (data > node->data) {
    node->right = insert(node->right, data);
  }
  return node;
}

void BST::remove(int data) {
  root = remove(root, data);
}

BST::Node* BST::remove(Node* node, int data) {
  if (node == nullptr) return node;
  
  if (data < node->data) {
    node->left = remove(node->left, data);
  } else if (data > node->data) {
    node->right = remove(node->right, data);
  } else {
    if (node->left == nullptr) {
      Node* temp = node->right;
      delete node;
      return temp;
    } 
    else if (node->right == nullptr) {
      Node* temp = node->left;
      delete node;
      return temp;
    } 
    else {
      Node* temp = findMin(node->right);
      node->data = temp->data;
      node->right = remove(node->right, temp->data);
    }
  }
  return node;
}

BST::Node* BST::findMin(Node* node) {
  while (node && node->left) {
    node = node->left;
  }
  return node;
}

bool BST::search(int data) {
  return search(root, data);
}

bool BST::search(Node* node, int data) {
  if (node == nullptr) return false;
  if (data == node->data) return true;

  return data < node->data ? search(node->left, data) : search(node->right, data);
}

void BST::printInorder() {
  InorderTraversal(root);
}

void BST::InorderTraversal(Node* node) {
  if (node == nullptr) return;
  InorderTraversal(node->left);
  std::cout << node->data << " ";
  InorderTraversal(node->right);
}

File: DataStructureAndAlgorithm/DataStructures/Trees/Python/binary_search_tree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None

    def insert(self, data):
        if not self.root:
            self.root = Node(data)
        else:
            self._insert(self.root, data)

    def _insert(self, node, data):
        if data < node.data:
            if node.left is None:
                node.left = Node(data)
            else:
                self._insert(node.left, data)
        elif data > node.data:
            if node.right is None:
                node.right = Node(data)
            else:
                self._insert(node.right, data)

    def remove(self, data):
        self.root = self._remove(self.root, data)

    def _remove(self, node, data):
        if node is None: return node

        if data < node.data:
            node.left = self._remove(node.left, data)
        elif data > node.data:
            node.right = self._remove(node.right, data)
        else:
            if node.left is None:
                return node.right
            elif node.right is None:
                return node.left
            
            min_larger_node = self._find_min(node.right)
            node.data = min_larger_node.data
            node.right = self._remove(node.right, min_larger_node.data)
        return node

    def _find_min(self, node):
        current = node
        while current.left is not None:
            current = current.left
        return current

    def search(self, data):
        return self._search(self.root, data)

    def _search(self, node, data):
        if node is None or node.data == data:
            return node is not None
        if data < node.data:
            return self._search(node.left, data)
        return self._search(node.right, data)

    def print_inorder(self):
        self._inorder_traversal(self.root)
        print()

    def _inorder_traversal(self, node):
        if node:
            self._inorder_traversal(node.left)
            print(node.data, end=" ")
            self._inorder_traversal(node.right)
